File: 026/026.map.py
# Not the best approach if asked the result itself.
# Can be done more straightforwards actually.
# Still, it was interesting to write (the ratio -> the decimal) converter.

def decimal(numerator: int, denominator: int, limit = None, get_period = False):
  '''
  converts ratio to it's decimal form.
    Args:
      limit: how many decimal digits to get.
              default value: None -> get full decimal. 
      get_period: returns period of ratio instead if needed (True)
    Returns:
      str   -> if the decimal is reccuring: '1.02(345)',
      float -> if the decimal is non-reccuring: 1.25,
      str   -> if get_period = True and the ratio has a period: '0235',
               if no period is present, returns '0'
  '''
  unit = numerator // denominator
  divident = 10 * (numerator - unit * denominator)
  # long division:
  #   intermediate divident: index of the string
  #   first state index begins from 2 (after zero and dot)
  #   same divident means repetition  
  #   from the last index that divident was encountered at
  states = {}
  decimal = f'{unit}.'
  # while:
  #   divident is present
  #   limit is not required OR still not met
  while divident and (limit == None or limit):
    previous_index = states.get(divident, None)
    if previous_index != None:
      period_start = previous_index
      non_periodic = decimal[ : period_start]
      period       = decimal[period_start : ]
      if get_period: return f'{period}'
      return f'{non_periodic}({period})'

    states[divident] = len(decimal)
    unit = divident // denominator
    decimal += str(unit)
    divident = 10 * (divident - unit * denominator)
    if limit: limit -= 1
  
  # if divident > 0: return float(decimal)  # limited decimal
  if get_period: return '0'                   # no period -> 0
  return float(decimal)

# decorator -> gives human answer
def informative(func):
  def func_wrapper(limit):
    denominator, period = func(limit)
    return f'The longest period length {period} has the fraction 1/{denominator}.'
  return func_wrapper

# search
@informative
def longest_period(limit):
  """
  Searches for the denominator with the longest period of the form 1/d.
    Args:
      limit: defines the search space, where denominator is in [1, limit)
    Returns
      A fraction and ith longest period: (denominator, period_length).
  """
  # dict -> denominator: length of the period for 1 / denominator
  periods = { d: len(decimal(1, d, get_period = True)) for d in range(1, limit) }
  denominator = max(periods, key = periods.get)
  return denominator, periods[denominator]

# results
print(longest_period(1000))

File: 026/026.map.js
// Not the best approach if asked the result itself.
// Can be done more straightforwards actually.
// Still, it was interesting to write (the ratio -> the decimal) converter.

function toDecimal(numerator, denominator, limit = null, getPeriod = false) {
  /**
   * Converts the given ratio to decimal form.
   * @param {int} numerator - the 1st member of the ratio.
   * @param {int} denominator - the 2nd member of the ratio.
   * @param {int} limit - limits the number of decimal digits to get,
   *                      default value: null -> returns the full decimal.
   * @param {boolean} getPeriod - if 'true' -> returns the period of the ratio instead.
   *                              default value: 'false'.
   * @returns {string} - if the decimal is reccuring -> '1.02(35)',
   * @returns {float}  - if the decimal is non-reccuring -> 2.5,
   * @returns {string} - if getPeriod = true, returns period if the ratio,
   *                     no period -> '0' is returned.   
   */
  let unit = ~ ~ (numerator / denominator);
  let divident = 10 * (numerator - unit * denominator);
  let decimal = `${unit}.`;
  // long division, searching for the same divident
  // stores information as {divident: index}
  // so we will be able to spot a period
  states = {};
  while (divident && limit || limit == null)
  {
    let previousIndex = states[divident]
    if (previousIndex) {
      let repetendStart = previousIndex;
      let nonRepeating = decimal.slice(0, repetendStart);
      let period = decimal.slice(repetendStart);
      if (getPeriod) {
        // return period if asked
        return period;
      }
      // reccuring decimal -> return
      return `${nonRepeating}(${period})`   
    }
    states[divident] = decimal.length;
    unit = ~ ~ (divident / denominator);
    decimal += unit.toString();
    divident = 10 * (divident - unit * denominator);
    if (limit) { limit--; };
    if (limit === 0) { break; }
  }
  // period is asked, but the ratio has not one
  if (getPeriod) { return '0'; }
  // the decimal is non-reccuring, return float
  return Number(decimal);
}

// solving the problem
search = (denominatorLimit) => {
  // Searches for the longest period of the fractions of form 1/denominator,
  //  where denominator is in [1, denominatorLimit)
  let [number, period] = [0, 0];
  for (let denominator = 1; denominator < denominatorLimit; denominator++)
  {
    let currentPeriod = toDecimal(1, denominator, undefined, true).length;
    if (currentPeriod > period) {
      [number, period] = [denominator, currentPeriod];
    }
  }
  return `The longest period length ${period} has denominator ${number}`;
};

// results
console.log(search(1000));